core/services/vrf/v2/listener_v2_log_processor.go

```diff
@@ -0,0 +1,1230 @@
+		nextTry(
+			req.attempts,
+			lsn.job.VRFSpec.BackoffInitialDelay,
+			lsn.job.VRFSpec.BackoffMaxDelay,
+			req.lastTry))
+}
+
+func nextTry(retries int, initial, max time.Duration, last time.Time) time.Time {
+	expBackoffFactor := math.Pow(backoffFactor, float64(retries-1))
+
+	var delay time.Duration
+	if expBackoffFactor > float64(max/initial) {
+		delay = max
+	} else {
+		delay = time.Duration(float64(initial) * expBackoffFactor)
+	}
+	return last.Add(delay)
+}
+
+// Remove all entries 10000 blocks or older
+// to avoid a memory leak.
+func (lsn *listenerV2) pruneConfirmedRequestCounts() {
+	min := lsn.blockNumberToReqID.FindMin()
+	for min != nil {
+		m := min.(fulfilledReqV2)
+		if m.blockNumber > (lsn.getLatestHead() - 10000) {
+			break
+		}
+		delete(lsn.respCount, m.reqID)
+		lsn.blockNumberToReqID.DeleteMin()
+		min = lsn.blockNumberToReqID.FindMin()
+	}
+}
+
+// Determine a set of logs that are confirmed
+// and the subscription has sufficient balance to fulfill,
+// given a eth call with the max gas price.
+// Note we have to consider the pending reqs already in the txm as already "spent" link or native,
+// using a max link or max native consumed in their metadata.
+// A user will need a minBalance capable of fulfilling a single req at the max gas price or nothing will happen.
+// This is acceptable as users can choose different keyhashes which have different max gas prices.
+// Other variables which can change the bill amount between our eth call simulation and tx execution:
+// - Link/eth price fluctuation
+// - Falling back to BHS
+// However the likelihood is vanishingly small as
+// 1) the window between simulation and tx execution is tiny.
+// 2) the max gas price provides a very large buffer most of the time.
+// Its easier to optimistically assume it will go though and in the rare case of a reversion
+// we simply retry TODO: follow up where if we see a fulfillment revert, return log to the queue.
+func (lsn *listenerV2) processPendingVRFRequests(ctx context.Context, pendingRequests []pendingRequest) {
+	confirmed := lsn.getConfirmedLogsBySub(lsn.getLatestHead(), pendingRequests)
+	var processedMu sync.Mutex
+	processed := make(map[string]struct{})
+	start := time.Now()
+
+	defer func() {
+		for _, subReqs := range confirmed {
+			for _, req := range subReqs {
+				if _, ok := processed[req.req.RequestID().String()]; ok {
+					// add to the inflight cache so that we don't re-process this request
+					lsn.inflightCache.Add(req.req.Raw())
+				}
+			}
+		}
+		lsn.l.Infow("Finished processing pending requests",
+			"totalProcessed", len(processed),
+			"totalFailed", len(pendingRequests)-len(processed),
+			"total", len(pendingRequests),
+			"time", time.Since(start).String(),
+			"inflightCacheSize", lsn.inflightCache.Size())
+	}()
+
+	if len(confirmed) == 0 {
+		lsn.l.Infow("No pending requests ready for processing")
+		return
+	}
+	var (
+		numWorkers int64 = 4
+		sem              = semaphore.NewWeighted(numWorkers)
```

Comment:
if we don't see performance bottle neck from tests, would it be better to remove parallelism here for simplicity?

Reply:
Yeah good note, I can remove it.

Reply:
Removed.